From 90ddede1137f49f3d432094be32c1b9aafde99bc Mon Sep 17 00:00:00 2001
From: Yongjun Zhang <yzhang@cloudera.com>
Date: Wed, 27 Jan 2016 08:04:25 -0800
Subject: [PATCH 1131/2848] HADOOP-12718. Incorrect error message by fs -put
 local dir without permission. (John Zhuge via
 Yongjun Zhang)

(cherry picked from commit 97056c3355810a803f07baca89b89e2bf6bb7201)

Change-Id: I76571e036ba764e5ca4444bbbbea309bdd86c086
---
 .../org/apache/hadoop/fs/RawLocalFileSystem.java   |    5 +++++
 1 file changed, 5 insertions(+)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/RawLocalFileSystem.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/RawLocalFileSystem.java
index ca815a3..9f600b3 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/RawLocalFileSystem.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/RawLocalFileSystem.java
@@ -33,6 +33,7 @@
 import java.io.FileDescriptor;
 import java.net.URI;
 import java.nio.ByteBuffer;
+import java.nio.file.AccessDeniedException;
 import java.util.Arrays;
 import java.util.EnumSet;
 import java.util.StringTokenizer;
@@ -385,6 +386,10 @@ public boolean delete(Path p, boolean recursive) throws IOException {
 
     String[] names = localf.list();
     if (names == null) {
+      if (!localf.canRead()) {
+        throw new AccessDeniedException("cannot open directory " + f +
+            ": Permission denied");
+      }
       return null;
     }
     results = new FileStatus[names.length];
-- 
1.7.9.5

